### PR TITLE
Remove exactPath, triePath, prefix and regex method in Route api

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
As @anuraaga suggested, we can hide those four methods because the customers do not use them and they can give more complexity.

Modifications:
- Remove `exactPath`, `triePath`, `prefix` and `regex` in `Route`.
- Add `RoutePathType` which represents the path which was specified when a `Route` is created.
- Add `pathType()` and `paths()` to `Route` so that we are able to get the paths which we were getting using those four methods.
- Fail fast when a user does not specify a prefix path with `CompositeService`.

Result:
- Fewer exposed APIs

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpDocServicePlugin.java [hunk 1]
```diff
@@ -174,41 +167,55 @@ private static void addMethodInfo(Map<Class<?>, Set<MethodInfo>> methodInfos,
...
 
         builder.availableMimeTypes(availableMimeTypes(route));
         return builder.build();
     }
 
-    private static String endpointPathMapping(Route route) {
-        final Optional<String> exactPath = route.exactPath();
-        if (exactPath.isPresent()) {
-            return EXACT + exactPath.get();
+    private static String normalizedPathParam(Route route) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpDocServicePlugin.java]: `normalizePathParams()` ? (no D, pluralize 'param')

File: core/src/main/java/com/linecorp/armeria/server/PathMapping.java [hunk 2]
```diff
@@ -57,33 +57,27 @@
...
      */
-    Optional<String> exactPath();
+    RoutePathType pathType();
 
     /**
-     * Returns the prefix of this {@link PathMapping} if this is a {@link PrefixPathMapping} or a
-     * {@link PathMappingWithPrefix}, {@link Optional#empty()} otherwise.
+     * Returns the list of paths that this {@link Route} has. The paths are different according to the value
+     * of {@link #pathType()}. If the path type has a {@linkplain RoutePathType#hasTriePath() trie path},
+     * the first element is the path that represents the type and the second element is the trie path.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/PathMapping.java]: this method will return a two-element list whose first element is ... and the second element is ...

File: core/src/main/java/com/linecorp/armeria/server/PathParamMapping.java [hunk 3]
```diff
@@ -41,7 +41,7 @@
  *   <li>A set of path parameters declared in the path pattern</li>
  * </ul>
  */
-final class DefaultPathMapping extends AbstractPathMapping {
+final class PathParamMapping extends AbstractPathMapping {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/PathParamMapping.java]: How about `ParameterizedPathMapping`?

File: core/src/main/java/com/linecorp/armeria/server/RoutePathType.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * The prefix path type. e.g, "/", "/foo/"
+     */
+    PREFIX(true),
+
+    /**
+     * The path which contains path parameters. e.g, "/:", "/foo/:/bar/:"
+     */
+    PATH_PARAM(true),
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/RoutePathType.java]: `PARAMETERIZED`? Please feel free to suggest a better name.

File: core/src/main/java/com/linecorp/armeria/server/RegexWithPrefixMapping.java [hunk 5]
```diff
@@ -20,24 +20,29 @@
...
-import java.util.Optional;
+import java.util.List;
 import java.util.Set;
 
 import javax.annotation.Nullable;
 
-final class PathMappingWithPrefix extends AbstractPathMapping {
+import com.google.common.collect.ImmutableList;
+
+final class RegexWithPrefixMapping extends AbstractPathMapping {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/RegexWithPrefixMapping.java]: `RegexPathMappingWithPrefix`?

File: core/src/main/java/com/linecorp/armeria/server/RoutePathType.java (hunk 4 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/RoutePathType.java]: I can't think of any better name. `PARAMETERIZED` is fine for me. Thanks!

**Comment by minwoox** [comment]
All fixed. PTAL 🙇

File: core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpDocServicePlugin.java [hunk 6]
```diff
@@ -174,41 +167,55 @@ private static void addMethodInfo(Map<Class<?>, Set<MethodInfo>> methodInfos,
...
-            return null;
-        }
-
-        final EndpointInfoBuilder builder = new EndpointInfoBuilder(hostnamePattern, endpointPathMapping);
-        if (endpointPathMapping.startsWith(REGEX) && route.prefix().isPresent()) {
-            // PathMappingWithPrefix
-            builder.regexPathPrefix(PREFIX + route.prefix().get());
+        final EndpointInfoBuilder builder;
+        final RoutePathType pathType = route.pathType();
+        final List<String> path = route.paths();
```
Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpDocServicePlugin.java]: Perhaps `pathComponents`

File: core/src/main/java/com/linecorp/armeria/server/CatchAllPathMapping.java [hunk 7]
```diff
@@ -18,17 +18,17 @@
...
+import java.util.List;
 import java.util.Set;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 
 final class CatchAllPathMapping extends AbstractPathMapping {
 
-    static final CatchAllPathMapping INSTANCE = new CatchAllPathMapping();
+    static final CatchAllPathMapping instance = new CatchAllPathMapping();
```
Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/server/CatchAllPathMapping.java]: Just to check, is this intentional? I think we've so far used all-caps for `INSTANCE`.

File: core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpDocServicePlugin.java (hunk 6 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpDocServicePlugin.java]: Sounds better. :-)

File: core/src/main/java/com/linecorp/armeria/server/CatchAllPathMapping.java (hunk 7 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/CatchAllPathMapping.java]: Yeah, we have used that. But it's technically not a constant. It's not public, so I changed it without much worry. Throughout the whole our code, we use two different conventions for the static final references. I want to change them all, but some of them are public so I couldn't.

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/CatchAllPathMapping.java]: Do you think we need to keep the variable using underscores with uppercase? 🤔

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/CatchAllPathMapping.java]: Changed it back to `INSTANCE`. :-)

**Comment by trustin** [comment]
Thanks, @minwoox !

**Comment by minwoox** [comment]
Thanks for reviewing. :wink:

**Pull request merged**